src/netbase.cpp

```diff
@@ -48,38 +48,57 @@ std::vector<CNetAddr> WrappedGetAddrInfo(const std::string& name, bool allow_loo
     ai_hint.ai_protocol = IPPROTO_TCP;
     // We don't care which address family (IPv4 or IPv6) is returned
     ai_hint.ai_family = AF_UNSPEC;
+
     // If we allow lookups of hostnames, use the AI_ADDRCONFIG flag to only
     // return addresses whose family we have an address configured for.
     //
     // If we don't allow lookups, then use the AI_NUMERICHOST flag for
     // getaddrinfo to only decode numerical network addresses and suppress
     // hostname lookups.
-    ai_hint.ai_flags = allow_lookup ? AI_ADDRCONFIG : AI_NUMERICHOST;
-
-    addrinfo* ai_res{nullptr};
-    const int n_err{getaddrinfo(name.c_str(), nullptr, &ai_hint, &ai_res)};
-    if (n_err != 0) {
-        return {};
-    }
-
-    // Traverse the linked list starting with ai_trav.
-    addrinfo* ai_trav{ai_res};
-    std::vector<CNetAddr> resolved_addresses;
-    while (ai_trav != nullptr) {
-        if (ai_trav->ai_family == AF_INET) {
-            assert(ai_trav->ai_addrlen >= sizeof(sockaddr_in));
-            resolved_addresses.emplace_back(reinterpret_cast<sockaddr_in*>(ai_trav->ai_addr)->sin_addr);
+    const int ai_flags = allow_lookup ? AI_ADDRCONFIG : AI_NUMERICHOST;
+
+    // getaddrinfo is called twice and to prevent duplicates a set is used
+    // std::set used over a std::unordered_set as CNetAddr does not have a hash function
+    std::set<CNetAddr> resolved_addresses_set;
+
+    auto callgetaddrinfo = [&](int flags, bool only_add_local_addr) {
+        ai_hint.ai_flags = flags;
+        addrinfo* ai_res{nullptr};
+        const int n_err{getaddrinfo(name.c_str(), nullptr, &ai_hint, &ai_res)};
+        if (n_err != 0) {
+            return;
```

Comment:
You could check specifically for `EAI_FAMILY` and then only recheck without `AI_ADDRCONFIG` in that condition. I think the code you have now will always check everything twice, but ignore duplicates?

Reply:
This makes sense to me as well and it might make for a significantly smaller code change.

Reply:
We still need to check again if we do get the error though so checking the error would be more performant (not always having to call `getaddrinfo` twice) but _more_ code.

Reply:
I looked into catching specifically `EAI_ADDRFAMILY` errors but this only applies on GNU Libc and is not POSIX. I checked windows and they have different errors. Rather than try to catalogue every possible system and libc implementation, the code now retries on any error (assuming you first tried with `AI_ADDRCONFIG` in the first place).